### PR TITLE
added Compact/Default/Minimal JsonSerializers with tests and rev'ed v…

### PR DESCRIPTION
**Pull request by SurajGupta**
…ersion

### CODE REVIEW

File: Recipes/JsonConfiguration/CompactJsonSerializer.cs [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+#endif
+    internal static partial class CompactJsonSerializer
+    {
+        /// <summary>
+        /// Deserializes the JSON to the specified .NET type.
+        /// </summary>
+        /// <typeparam name="T">The type of the object to deserialize to.</typeparam>
+        /// <param name="value">The JSON to deserialize.</param>
+        /// <returns>The deserialized object from the JSON string.</returns>
+        public static T DeserializeObject<T>(string value) where T : class
```
Comment (vrobinson) [inline Recipes/JsonConfiguration/CompactJsonSerializer.cs]: Thought you were going to drop "Object" from all these names? I thought the terser name was better.

File: Recipes/JsonConfiguration/CompactJsonSerializer.cs [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        {
+            var result = JsonConvert.SerializeObject(value, JsonConfiguration.CompactSerializerSettings);
+            return result;
+        }
+
+        /// <summary>
+        /// Serializes the specified object to a JSON string using a type.
+        /// </summary>
+        /// <param name="value">The object to serialize.</param>
+        /// <param name="type">The type of the value being serialized. This parameter is used when TypeNameHandling is Auto to write out the type name if the type of the value does not match. Specifing the type is optional.</param>
```
Comment (vrobinson) [inline Recipes/JsonConfiguration/CompactJsonSerializer.cs]: Do we specify what TypeNameHandling is? If not, what is the default? Because depending upon what we pick for a particular serializer settings and how this method would behave under that setting, it may make sense to drop this method from some or all of these types (or maybe not).

File: Recipes.Test/JsonConfiguration/CompactJsonSerializerTest.cs [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+            Assert.That(json, Is.EqualTo("{\"name\":\"spud\",\"furColor\":\"brindle\",\"dogTag\":null,\"age\":5}"));
+        }
+
+        [Test]
+        public void SerializeObject_with_type_serializes_to_json_using_CompactSerializerSettings()
+        {
+            var dog = new Dog(5, "spud", FurColor.Brindle);
+
+            var json = CompactJsonSerializer.SerializeObject(dog, typeof(Animal));
```
Comment (vrobinson) [inline Recipes.Test/JsonConfiguration/CompactJsonSerializerTest.cs]: I thought this embedded the typename in the json? If not, what is the difference in this method overload? This is really the same question I asked in the object's source.

File: Recipes.Test/JsonConfiguration/CompactJsonSerializerTest.cs [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+
+            Assert.That(json, Is.EqualTo("{\"name\":\"spud\",\"furColor\":\"brindle\",\"dogTag\":null,\"age\":5}"));
+        }
+
+        [Test]
+        public void DeserializeObject_generic_deserializes_json_using_CompactSerializerSettings()
```
Comment (vrobinson) [inline Recipes.Test/JsonConfiguration/CompactJsonSerializerTest.cs]: No biggie.... I usually identify a generic method in a test name by calling it something like DeserializeObjectOfT_does_whatever

File: Recipes.Test/JsonConfiguration/CompactJsonSerializerTest.cs [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+        {
+            var dogJson = "{\"name\":\"Barney\",\"furColor\":\"brindle\",\"age\":10}";
```
Comment (vrobinson) [inline Recipes.Test/JsonConfiguration/CompactJsonSerializerTest.cs]: Because it took compact input it deserialized with the CompactSerializerSettings? Is there anything a bit more concrete/unique about the deserialization that can be tested?

File: Recipes.Test/JsonConfiguration/CompactJsonSerializerTest.cs [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+
+            var dog = CompactJsonSerializer.DeserializeObject<Animal>(dogJson) as Dog;
+
+            Assert.That(dog, Is.Not.Null);
```
Comment (vrobinson) [inline Recipes.Test/JsonConfiguration/CompactJsonSerializerTest.cs]: Similarly, are we testing that it can deserialize the type? Or that it is using CompactSerializerSettings?

File: Recipes.Test/JsonConfiguration/CompactJsonSerializerTest.cs [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+
+            Assert.That(json, Is.EqualTo("{\"name\":\"spud\",\"furColor\":\"brindle\",\"dogTag\":null,\"age\":5}"));
+        }
+
+        [Test]
+        public void DeserializeObject_generic_deserializes_json_using_CompactSerializerSettings()
+        {
+            var dogJson = "{\"name\":\"Barney\",\"furColor\":\"brindle\",\"age\":10}";
+
+            var dog = CompactJsonSerializer.DeserializeObject<Animal>(dogJson) as Dog;
```
Comment (vrobinson) [inline Recipes.Test/JsonConfiguration/CompactJsonSerializerTest.cs]: Wouldn't I normally call DeserializeObject<Dog>()? Or is this the unique thing about Compact you are using to test here?

File: Recipes.Test/JsonConfiguration/CompactJsonSerializerTest.cs [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+            Assert.That(dog.FurColor, Is.EqualTo(FurColor.Brindle));
+            Assert.That(dog.DogTag, Is.Null);
+        }
+
+        [Test]
+        public void DeserializeObject_without_type_deserializes_json_into_JObject_using_CompactSerializerSettings()
+        {
+            var dogJson = "{\"name\":\"Barney\",\"furColor\":\"brindle\",\"age\":10}";
+
+            var dog = CompactJsonSerializer.DeserializeObject(dogJson) as JObject;
```
Comment (vrobinson) [inline Recipes.Test/JsonConfiguration/CompactJsonSerializerTest.cs]: I assume the source is setup with the same signature as the underlying JsonConvert it is encapsulating? I was anticipating the return could be cast to an Animal or Dog until I saw the JObject.

File: Recipes.Test/JsonConfiguration/CompactJsonSerializerTest.cs [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+
+            Assert.That(dog.Properties().Count(), Is.EqualTo(3));
+            Assert.That(dog["name"].ToString(), Is.EqualTo("Barney"));
+            Assert.That(dog["age"].ToString(), Is.EqualTo("10"));
+            Assert.That(dog["furColor"].ToString(), Is.EqualTo("brindle"));
+        }
+
+        [Bindable(true)]
+        private class Animal
```
Comment (vrobinson) [inline Recipes.Test/JsonConfiguration/CompactJsonSerializerTest.cs]: Since these get used in multiple places how about elevating to internal classes from private ones so they can be shared across test cases?

File: Recipes.Test/JsonConfiguration/CompactJsonSerializerTest.cs [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+        private enum FurColor
+        {
+            Black,
+
+            Brindle,
+
+            Golden
+        }
+
+        [SuppressMessage("Microsoft.Performance", "CA1812:AvoidUninstantiatedInternalClasses", Justification = "Class is used but is constructed via reflection and code analysis cannot detect that.")]
```
Comment (vrobinson) [inline Recipes.Test/JsonConfiguration/CompactJsonSerializerTest.cs]: Perhaps elevating the classes will allow you to remove some of the suppressions...

File: Recipes.Test/JsonConfiguration/DefaultJsonSerializerTest.cs [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+
+            var json = DefaultJsonSerializer.SerializeObject(dog, typeof(Animal));
+
+            Assert.That(json, Is.EqualTo("{\r\n  \"name\": \"spud\",\r\n  \"furColor\": \"brindle\",\r\n  \"dogTag\": null,\r\n  \"age\": 5\r\n}"));
+        }
+
+        [Test]
+        public void DeserializeObject_generic_deserializes_json_using_DefaultSerializerSettings()
+        {
+            var dogJson = "{\"name\":\"Barney\",\"furColor\":\"brindle\",\"age\":10}";
```
Comment (vrobinson) [inline Recipes.Test/JsonConfiguration/DefaultJsonSerializerTest.cs]: This looks more like Compact then Default data. What are we trying to test here (see feedback on CompactJsonSerializerTest for more context).

File: Recipes.Test/JsonConfiguration/MinimalJsonSerializerTest.cs [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+
+    using NUnit.Framework;
+
+    [TestFixture]
+    public class MinimalJsonSerializerTest
+    {
+        [Test]
+        public void SerializeObject_without_type_serializes_to_json_using_MinimalSerializerSettings()
+        {
+            var dog = new Dog(5, "spud", FurColor.Brindle);
```
Comment (vrobinson) [inline Recipes.Test/JsonConfiguration/MinimalJsonSerializerTest.cs]: Maybe something needs to be null and see that get stripped from the output to verify this is using Minimal.

File: Recipes.Test/JsonConfiguration/CompactJsonSerializerTest.cs (hunk 3 above)
Comment (SurajGupta) [inline Recipes.Test/JsonConfiguration/CompactJsonSerializerTest.cs]: good point - right now TypeNameHandling == None for all serialization settings, so this method overload does not add any functionality

Comment (SurajGupta) [inline Recipes.Test/JsonConfiguration/CompactJsonSerializerTest.cs]: I'm removing the overload + tests

File: Recipes.Test/JsonConfiguration/CompactJsonSerializerTest.cs (hunk 4 above)
Comment (SurajGupta) [inline Recipes.Test/JsonConfiguration/CompactJsonSerializerTest.cs]: I'll fix - its good to be consistent about conventions

File: Recipes.Test/JsonConfiguration/CompactJsonSerializerTest.cs (hunk 5 above)
Comment (SurajGupta) [inline Recipes.Test/JsonConfiguration/CompactJsonSerializerTest.cs]: I'm not testing nor conditioning the test on how the input is structured, because this test is about deserialization, not serialization.  What I'm really testing here is that the CamelCaseStrictConstructor contract resolver is used.  I'm not sure what else I could easily test to uniqueness between Default/Compact/Minimal.  That said, your comment made me realize that I am testing this wrong.  I changed Cat to Mouse.  With Mouse and Dog, strict constructor matching must be used to deserialize.

File: Recipes.Test/JsonConfiguration/CompactJsonSerializerTest.cs (hunk 7 above)
Comment (SurajGupta) [inline Recipes.Test/JsonConfiguration/CompactJsonSerializerTest.cs]: Not sure what you mean - you would typically call DeserializeObject<T>.  See comment above.

File: Recipes.Test/JsonConfiguration/CompactJsonSerializerTest.cs (hunk 6 above)
Comment (SurajGupta) [inline Recipes.Test/JsonConfiguration/CompactJsonSerializerTest.cs]: We are testing that its using CompactSerializerSettings by testing the contract resolver.  While that doesn't differentiate it from DefaultSerializerSettings, it does ensure that we didn't use the Newtonsoft default, which I think is a good check.

File: Recipes.Test/JsonConfiguration/CompactJsonSerializerTest.cs (hunk 8 above)
Comment (SurajGupta) [inline Recipes.Test/JsonConfiguration/CompactJsonSerializerTest.cs]: Yes, the source is setup with the same signature as the underlying JsonConvert.  I too thought that I could cast to an Animal or Dog - but upon further reflection I'm not sure how that would work without Json doing some type-discovery-reflection.  If you spend a moment thinking about the implementation, then I think you'll see why JObject is a reasonable return type.

File: Recipes.Test/JsonConfiguration/CompactJsonSerializerTest.cs (hunk 9 above)
Comment (SurajGupta) [inline Recipes.Test/JsonConfiguration/CompactJsonSerializerTest.cs]: I like the idea - what should I call the file?  I'm going to put them in JsonSerializerTest for now and you can slap me later.

Comment (SurajGupta) [inline Recipes.Test/JsonConfiguration/CompactJsonSerializerTest.cs]: created Animal.cs and Lighting.cs - unfortuantely that requires more code analysis suppressions

File: Recipes.Test/JsonConfiguration/CompactJsonSerializerTest.cs (hunk 10 above)
Comment (SurajGupta) [inline Recipes.Test/JsonConfiguration/CompactJsonSerializerTest.cs]: unfortunately not - see comment above.

Comment (SurajGupta) [inline Recipes.Test/JsonConfiguration/CompactJsonSerializerTest.cs]: well...I think I was able to remove some, but some were added.  I basically started with a blank slate (no suppressions) and build-up from there.

File: Recipes.Test/JsonConfiguration/DefaultJsonSerializerTest.cs (hunk 11 above)
Comment (SurajGupta) [inline Recipes.Test/JsonConfiguration/DefaultJsonSerializerTest.cs]: see comments above - i'm testing deserialization

File: Recipes.Test/JsonConfiguration/MinimalJsonSerializerTest.cs (hunk 12 above)
Comment (SurajGupta) [inline Recipes.Test/JsonConfiguration/MinimalJsonSerializerTest.cs]: I am verifying minimal because the null Nickname (new property added to the Dog type) is not serialized

File: Recipes/JsonConfiguration/CompactJsonSerializer.cs (hunk 1 above)
Comment (SurajGupta) [inline Recipes/JsonConfiguration/CompactJsonSerializer.cs]: I liked the terser name too, but if you look at the list of JsonConvert methods, you'll see DeserializeAnonymousType.  So if we add that later, I thought I might be confusing so I made it consistent with JsonConvert and used DeserializeObject.  Same issue with Serialize (there's a SerializeXmlNode and others)

http://www.newtonsoft.com/json/help/html/Methods_T_Newtonsoft_Json_JsonConvert.htm

File: Recipes/JsonConfiguration/CompactJsonSerializer.cs (hunk 2 above)
Comment (SurajGupta) [inline Recipes/JsonConfiguration/CompactJsonSerializer.cs]: I removed this overload per the above comment.

File: Recipes.Test/JsonConfiguration/CompactJsonSerializerTest.cs (hunk 5 above)
Comment (vrobinson) [inline Recipes.Test/JsonConfiguration/CompactJsonSerializerTest.cs]: The fix was changing Cat to Mouse - LOL!

File: Recipes.Test/JsonConfiguration/CompactJsonSerializerTest.cs (hunk 7 above)
Comment (vrobinson) [inline Recipes.Test/JsonConfiguration/CompactJsonSerializerTest.cs]: My bad - meant to say DeserializeObject<Dog>()

Comment (vrobinson) [inline Recipes.Test/JsonConfiguration/CompactJsonSerializerTest.cs]: mmm... not my bad - it's gihub removing what it thinks is an html tag - let me try again: DeserializeObject &lt;Dog&gt;()

File: Recipes.Test/JsonConfiguration/CompactJsonSerializerTest.cs (hunk 10 above)
Comment (vrobinson) [inline Recipes.Test/JsonConfiguration/CompactJsonSerializerTest.cs]: Sounds good

File: Recipes.Test/JsonConfiguration/DefaultJsonSerializerTest.cs (hunk 11 above)
Comment (vrobinson) [inline Recipes.Test/JsonConfiguration/DefaultJsonSerializerTest.cs]: Do you see my point that the json blob you are deserializing looks more like Compact than Default?

File: Recipes.Test/JsonConfiguration/MinimalJsonSerializerTest.cs (hunk 12 above)
Comment (vrobinson) [inline Recipes.Test/JsonConfiguration/MinimalJsonSerializerTest.cs]: Maybe put a comment in to note this then.

File: Recipes.Test/JsonConfiguration/CompactJsonSerializerTest.cs (hunk 7 above)
Comment (SurajGupta) [inline Recipes.Test/JsonConfiguration/CompactJsonSerializerTest.cs]: yeah,  same issue here.  I meant to say what you said.  SO I think we're good or is there still a question?

File: Recipes.Test/JsonConfiguration/DefaultJsonSerializerTest.cs (hunk 11 above)
Comment (SurajGupta) [inline Recipes.Test/JsonConfiguration/DefaultJsonSerializerTest.cs]: Yup - totally see your point, I'm just saying that it doesn't matter.  It was easier to encode this as Compact because I didn't have to do the newlines.

File: Recipes.Test/JsonConfiguration/MinimalJsonSerializerTest.cs (hunk 12 above)
Comment (SurajGupta) [inline Recipes.Test/JsonConfiguration/MinimalJsonSerializerTest.cs]: yeah good call

File: Recipes/JsonConfiguration/JsonConfiguration.nuspec [hunk 13]
```diff
@@ -17,8 +17,8 @@
         <dependencies>
             <group targetFramework="net450">
                 <dependency id="Spritely.Recipes.SecureStringJsonConverter" version="0.3.1" />
-                <dependency id="Spritely.Recipes.InheritedTypeJsonConverter" version="0.7.0" />
-                <dependency id="Spritely.Recipes.CamelStrictConstructorContractResolver" version="0.1.0" />
+                <dependency id="Spritely.Recipes.InheritedTypeJsonConverter" version="0.7.1" />
```
Comment (vrobinson) [inline Recipes/JsonConfiguration/JsonConfiguration.nuspec]: You need to rev CamelStrict.... in InheritedTypeJsonConverter.nuspec too.

Comment (vrobinson) [inline Recipes/JsonConfiguration/JsonConfiguration.nuspec]: My bad - these aren't dependent - ignore
